spinalcordtoolbox/moco/dl/inference.py

```diff
@@ -0,0 +1,211 @@
+        bvecs (dmri only):     {fname_bvecs if fname_bvecs else 'None'}
+        bvals (dmri only):     {fname_bvals if fname_bvals else 'None'}
+    """))
+
+    # Load image
+    mov_img = Image(fname_data)
+    mov_img.change_orientation('RPI')
+    mov_np = mov_img.data.astype(np.float32)
+    affine, header = mov_img.affine, mov_img.header
+
+    mask_img = Image(fname_mask)
+    mask_img.change_orientation('RPI')
+    mask_np = mask_img.data.astype(np.float32)
+
+    if fname_ref:
+        ref_img = Image(fname_ref)
+        ref_img.change_orientation('RPI')
+        ref_np = ref_img.data.astype(np.float32)
+    else:
+        ref_np = mov_np[..., 0]
+
+    # Load DL model
+    print(f"[moco-dl] Loading checkpoint: {ckpt_path}")
+    model = DenseRigidReg()
+    state = torch.load(ckpt_path, map_location=device)
+    model.load_state_dict(state["state_dict"])
+    model = model.to(device)
+    model.eval()
+    model.warp = RigidWarp(mode="bilinear")
+
+    moving = torch.from_numpy(mov_np).unsqueeze(0).unsqueeze(0).to(device)
+    fixed = torch.from_numpy(ref_np).unsqueeze(0).unsqueeze(0).to(device)
+    mask = torch.from_numpy(mask_np).unsqueeze(0).unsqueeze(0).to(device)
+
+    # Run inference
+    start_time = time.time()
+    print("[moco-dl] Starting inference...")
+    with torch.no_grad():
+        warped, flow, Tx, Ty = model(moving, fixed, mask)
+
+    inference_time = time.time() - start_time
+    print(f"Inference time: {inference_time:.2f} sec")
+
+    warped = warped.squeeze().cpu().numpy()  # (H,W,D,T)
+    flow = flow.squeeze().cpu().numpy()
+    Tx = Tx.squeeze().cpu().numpy()  # (D,T)
+    Ty = Ty.squeeze().cpu().numpy()  # (D,T)
+
+    H, W, D, T = warped.shape
+    # Restore high-frequency detail from raw data
+    sharpened = np.copy(warped)
+    for t in range(T):
+        for d in range(D):
+            img_warped = warped[..., d, t]
+            img_raw = mov_np[..., d, t]
+            mask_slice = mask_np[..., d]
+
+            if np.count_nonzero(mask_slice) == 0:
+                sharpened[..., d, t] = img_warped
+                continue
+
+            raw_smooth = smooth(img_raw.astype(np.float32), sigmas=[0.5, 0.5])
+            texture = img_raw - raw_smooth
+            out = img_warped + 1.2 * texture
+            lo, hi = np.percentile(img_raw[mask_slice > 0], [0.5, 99.5])
+            out = np.clip(out, lo, hi)
+            sharpened[..., d, t] = out
+
+    matched = np.zeros_like(sharpened)
+    for t in range(T):
+        matched[..., t] = ski_exposure.match_histograms(
+            sharpened[..., t].astype(np.float32),
+            mov_np[..., t].astype(np.float32)
+        )
+
+    # Save Moco output
+    tmp_main = os.path.join(ofolder, "tmp_mocoDL.nii.gz")
+    im_main = Image(matched, hdr=header)
+    im_main.affine = affine
+    im_main.save(tmp_main)
+
+    base_output = add_suffix(fname_data, "_mocoDL")
+    fname_moco = generate_output_file(tmp_main, os.path.join(ofolder, base_output))
```

Comment:
Similarly, I would recommend not using `generate_output_file` for the intermediate tempfiles. You can just use `im.save()` for the tempfiles.